src/SmurfBuilder.cpp

```diff
@@ -82,106 +77,178 @@ void SmurfBuilder::FlushStash(){
     int sample = 0;
-    for (auto x = read_stash_.begin(); x != read_stash_.end(); x++, sample++){
-        auto hdr = (*x)->sp->getHeader();
-
-        sample_times[sample] = (*x)->GetTime();
-
-        (*primary_vec[0])[sample] = hdr->getUnixTime();
-        (*primary_vec[1])[sample] = hdr->getFluxRampIncrement();
-        (*primary_vec[2])[sample] = hdr->getFluxRampOffset();
-        (*primary_vec[3])[sample] = hdr->getCounter0();
-        (*primary_vec[4])[sample] = hdr->getCounter1();
-        (*primary_vec[5])[sample] = hdr->getCounter2();
-        (*primary_vec[6])[sample] = hdr->getAveragingResetBits();
-        (*primary_vec[7])[sample] = hdr->getFrameCounter();
-        (*primary_vec[8])[sample] = hdr->getTESRelaySetting();
+    for (auto it = start; it != stop; it++, sample++){
+        sample_times[sample] = (*it)->GetTime();
+
+        auto hdr = (*it)->sp->getHeader();
+        primary_buffer[sample + 0 * nsamps] = hdr->getUnixTime();
+        primary_buffer[sample + 1 * nsamps] = hdr->getFluxRampIncrement();
+        primary_buffer[sample + 2 * nsamps] = hdr->getFluxRampOffset();
+        primary_buffer[sample + 3 * nsamps] = hdr->getCounter0();
+        primary_buffer[sample + 4 * nsamps] = hdr->getCounter1();
+        primary_buffer[sample + 5 * nsamps] = hdr->getCounter2();
+        primary_buffer[sample + 6 * nsamps] = hdr->getAveragingResetBits();
+        primary_buffer[sample + 7 * nsamps] = hdr->getFrameCounter();
+        primary_buffer[sample + 8 * nsamps] = hdr->getTESRelaySetting();
+
+        for (int i = 0; i < N_TES_BIAS; i++)
+            tes_buffer[sample + i * nsamps] = hdr->getTESBias(i);
+
+        for (int i = 0; i < nchans; i++){
+            data_buffer[sample + i * nsamps] = (*it)->sp->getData(i);
+        }
     }
 
-    primary_map->times = sample_times;
+    PyGILState_STATE gstate;
+    gstate = PyGILState_Ensure();
 
-    int i = 0;
-    for (auto key : primary_keys){
-        primary_map->insert(std::make_pair(key, primary_vec[i]));
-        i++;
-    }
+    data_ts->times = sample_times;
+    data_ts->SetDataFromBuffer((void*)data_buffer, 2, data_shape, NPY_INT32,
+            std::pair<int,int>(0, nsamps));
+    data_ts->Options(
+        enable_compression_, flac_level_, bz2_work_factor_, data_encode_algo_,
+        time_encode_algo_
+    );
+
+    tes_ts->times = G3VectorTime(sample_times);
+    tes_ts->SetDataFromBuffer((void*)tes_buffer, 2, tes_shape, NPY_INT32,
+            std::pair<int,int>(0, nsamps));
+    tes_ts->Options(
+        enable_compression_, flac_level_, bz2_work_factor_, data_encode_algo_,
+        time_encode_algo_
+    );
+
+    primary_ts->times = G3VectorTime(sample_times);
+    primary_ts->SetDataFromBuffer((void*)primary_buffer, 2, primary_shape,
+            NPY_INT64, std::pair<int,int>(0, nsamps));
+    primary_ts->Options(
+        enable_compression_, flac_level_, bz2_work_factor_, data_encode_algo_,
+        time_encode_algo_
+    );
```

Comment:
You are using data_encode_algo_, but I think you made special variables for the tes_bias and primary algos, right?

Reply:
Yep! Good catch